server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -282,46 +269,36 @@ private boolean prepareForUpload(Long generation) throws IOException {
                         logger.trace("Creating new writer for gen: [{}]", current.getGeneration() + 1);
                         current = createWriter(current.getGeneration() + 1);
                     }
+                    assert writeLock.isHeldByCurrentThread() : "Write lock must be held before we acquire the read lock";
+                    // Here we are downgrading the write lock by acquiring the read lock and releasing the write lock
+                    // This ensures that other threads can still acquire the read locks while also protecting the
+                    // readers and writer to not be mutated any further.
```

Comment:
Should we move this check below acquiring read lock to ensure the same thread also holds the write lock

Reply:
Are you suggesting to move the lock after acquiring the read lock?

Reply:
Sorry , I thought this was an already resolved conversation .

Reply:
@Bukhtawar - If it is a different thread than the one that has acquired the write lock, it is a dead lock.